File: Runtimes/CsMora/CsMora/Mora/Com/UDP/UDPConnections.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Mora.Com.UDP
{


    public class UDPConnection : AbstractConnection
    {
        private Socket mSocket;
        public UDPConnection(RemoteCommunicator com, int idx) : base(idx)
        {
            mSocket = new Socket(com.Address.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
            mSocket.Connect(com.Address, com.Port);
        }

        public override int getMaximumMessageSize()
        {
            return 64000;
        }

        public override bool Send(OutMsg msg)
        {
            int l = mSocket.Send(msg.data);
            return l == msg.data.Length;
        }

        internal override void Close()
        {
            mSocket.Close();
            mSocket = null;
        }

    }

    public class UDPServer : IServer
    {
        private bool mAlive = false;
        private Thread mServerThread;
        public static ManualResetEvent allDone = new ManualResetEvent(false);
        private Communicator mCommunicator;
        private IPEndPoint mEndpoint;
        private RemoteCommunicator mResponseCommunicator;


        public void StartListening(Communicator communicator)
        {
            if (mAlive == true)
            {
                throw new Exception("Server already started");
            }
            mAlive = true;
            mCommunicator = communicator;
            mServerThread = new Thread(new ThreadStart(this.runServer));
            mServerThread.Start();
        }

        public void StopListening()
        {
            mAlive = false;
            if (mServerThread != null)
            {
                allDone.Set();
                //send some null-message to the socket
                mServerThread.Join();
                mServerThread = null;
            }
        }

        public IPEndPoint LocalEndpoint()
        {
            return mEndpoint;
        }
        public RemoteCommunicator GetResponseCommunicator()
        {
            if (mResponseCommunicator == null && LocalEndpoint() != null)
            {
                mResponseCommunicator = new RemoteCommunicator(LocalEndpoint().Address, LocalEndpoint().Port, PROTOCOL.UDP);
            }
            return mResponseCommunicator;
        }






        private void runServer()
        {
            Options opt = mCommunicator.Options;

            
            // Create a TCP/IP socket.  
            //Socket socketListener = new Socket(opt.Adr.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
            int port = opt.Port;
            if (port < 0)
                port = 0;
            IPEndPoint localEndPoint = new IPEndPoint(opt.Adr, opt.Port);
            try
            {
                UdpClient listener = new UdpClient(localEndPoint);
                mEndpoint = listener.Client.LocalEndPoint as IPEndPoint;
                IPEndPoint remoteEP = new IPEndPoint(IPAddress.Any, port);
                int bytesRead = 0;
                InMsg msg = null;
                int state = 0;
                while (mAlive)
                {
                    byte[] buffer = listener.Receive(ref remoteEP);
                    if (buffer.Length > 0)
                    {
                        bytesRead = buffer.Length;
                        switch (state)
                        {
                            case 0:
                                msg = new InMsg(buffer, bytesRead);
                                bytesRead = 0;
                                break;
                            case 1:
                                byte[] remaining = msg.Append(buffer, bytesRead);
                                if (remaining != null)
                                {
                                    Array.Copy(remaining, 0, buffer, 0, remaining.Length);
                                    bytesRead = remaining.Length;
                                }
                                else
                                    bytesRead = 0;
                                break;
                        }
                        if (msg.complete())
                        {
                            handleNetMessage(msg, remoteEP);
                            if (msg.RequestsAck())
                            {
                                byte[] ackMsg = msg.AckMessage();
                            }
                            state = 0;
                            msg = null;
                        }
                        else
                            state = 1;
                    }
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e.ToString());
            }
        }

        private Dictionary<byte, InMessageCollection> mCollections = new Dictionary<byte, InMessageCollection>();
        private void handleNetMessage(InMsg msg, EndPoint remoteEndPoint)
        {
            InMessageCollection mc;
            if (mCollections.ContainsKey(msg.messageMagic) == false)
            {
                mCollections.Add(msg.messageMagic, mc = new InMessageCollection(msg, remoteEndPoint));
            }
            else
            {
                mc = mCollections[msg.messageMagic];
                mc.Append(msg);
            }
            if (mc.IsComplete())
            {
                mCollections.Remove(msg.messageMagic);
                mc.BuildStream();
                mCommunicator.HandleIncomingMessage(mc);
            }
        }

       
    }
}

File: Runtimes/JMora/src/main/java/de/sos/mora/com/UDPConnections.java
package de.sos.mora.com;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.HashMap;

import de.sos.mora.RemoteObjects.RemoteCommunicator;
import de.sos.mora.com.ClientConnections.AbstractConnection;
import de.sos.mora.com.Communicator.Options;
import de.sos.mora.com.Communicator.PROTOCOL;
import de.sos.mora.com.NetMessagees.InMessageCollection;
import de.sos.mora.com.NetMessagees.InMsg;
import de.sos.mora.com.NetMessagees.OutMsg;

public class UDPConnections implements ISocketServer{

	public static class UDPConnection extends AbstractConnection {
		
		
		private RemoteCommunicator mTarget;
		private DatagramSocket mServerSocket;
		private DatagramPacket mPacket;

		public UDPConnection(RemoteCommunicator target, int index) throws IOException {
			super(index);
			mTarget = target;
			mServerSocket = new DatagramSocket();
			mPacket = new DatagramPacket(new byte[] {0}, 1, mTarget.address, mTarget.port);
		}

		
		@Override
		public int getMaximumMessageSize() {
			return 64000;
		}
		@Override
		protected boolean send(OutMsg outMsg) {
			try {				
				mPacket.setData(outMsg.getBuffer());
				mServerSocket.send(mPacket);
			} catch (IOException e) {
				e.printStackTrace();
				return false;
			}
			return true;
		}
		
		@Override
		public void close() {
			try {
				mServerSocket.close();
			} catch (Exception e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
	}
	
	
	private class ServerRunnable implements Runnable{
		public int 			mUsedPort;
		public InetAddress	mUsedHost;
		
		private static final int NEW_MESSAGE = 0;
		private static final int CONTINUE_MESSAGE = 1;
		
		
		DatagramSocket		mSocketServer;
		private boolean mAlive;
		public void run() {
			try {
				mSocketServer = new DatagramSocket(mOptions.port < 0 ? 0 : mOptions.port, mOptions.adr);
								
				
				mUsedHost = mSocketServer.getLocalAddress();
				mUsedPort = mSocketServer.getLocalPort();
				mResponseAddress = new RemoteCommunicator(mUsedHost, mUsedPort, PROTOCOL.UDP);
				System.out.println("Start UDP Server at: " + mResponseAddress);

				mAlive = true;
				byte[] receiveData = new byte[64000]; //64000 == theoretical maximum size for an udp socket package				
				while(mAlive = true) {
					int bytesRead = 0;
					int state = NEW_MESSAGE;
					InMsg msg = null;
					InetAddress senderIP = null;
					while(bytesRead <= NetMessagees.HEADER_LENGTH) {
						DatagramPacket receivedPacket = new DatagramPacket(receiveData, bytesRead, receiveData.length);
						mSocketServer.receive(receivedPacket);
						
						bytesRead += receivedPacket.getLength();
						senderIP = receivedPacket.getAddress();
					}
					if (bytesRead > 0) {
						switch(state) {
						case NEW_MESSAGE: {
							msg = new InMsg(receiveData, bytesRead);
							bytesRead = 0;
							break;
						}
						case CONTINUE_MESSAGE:{
							byte[] remaining = msg.append(receiveData, bytesRead);
							if (remaining != null) {
								System.arraycopy(remaining, 0, receiveData, 0, remaining.length);
								bytesRead = remaining.length;
							}else
								bytesRead = 0;
							break;
						}
						default:
							break;
						}
						
						if (msg.complete()) {
							handleNetMessage(msg, senderIP);
							state = NEW_MESSAGE;
						}else {
							state = CONTINUE_MESSAGE;
						}
					}
				}

			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}	
		}
		
		/** incompleted message collections, for the case that a message had to be splitted up into different messages, to fit size requirements*/
		private HashMap<Integer, InMessageCollection> 	mMessageCollections = new HashMap<>();
		
		private void handleNetMessage(InMsg msg, InetAddress senderIP) throws IOException {
			Integer key = Integer.valueOf(msg.messageMagic);
			InMessageCollection mc;
			if (mMessageCollections.containsKey(key)) {
				mc = mMessageCollections.get(key);
				mc.append(msg);
			}else {
				mMessageCollections.put(key, mc = new InMessageCollection(msg, senderIP));
			}
			
			if ((msg.flags & NetMessagees.REQUEST_ACKNOWLEDGE) == NetMessagees.REQUEST_ACKNOWLEDGE){
				//NetMessagees.acknowledgeMessage(mSocket, msg.messageMagic, msg.messageNumber);
				System.out.println("TODO: SEND ACKNOWLEDGE");
			}
			
			
			if (mc.isComplete()) {
				mc = mMessageCollections.remove(key);
				mc.complete();
				//This call will handle the message (may in another thread)
				mCommunicator.handleIncomingMessage(mc);
			}
		}
	}


	
	
	private final Options					mOptions;	
	private ServerRunnable					mServerRunnable = null;
	private Thread							mServerThread = null;
	private RemoteCommunicator				mResponseAddress;
	private Communicator 					mCommunicator;
	
	
	public UDPConnections(final Options opt) {
		mOptions = opt;
	}
	
	public void start(Communicator communicator) {
		mCommunicator = communicator;
		mServerThread = new Thread(mServerRunnable = new ServerRunnable());
		mServerThread.start();
	}
	
	public void waitForTermination() throws InterruptedException {
		if (mServerThread == null)
			return ;
		mServerThread.join();
	}


	public RemoteCommunicator getResponseAddress() {
		return mResponseAddress;
	}

	@Override
	public void stop() throws InterruptedException {
		mServerRunnable.mAlive = false;
		waitForTermination();
	}
	
}
